### ISSUE TITLE
DataFrame built from ndarray does not preserve `.values.dtype` if there are no columns

### ISSUE THREAD

**Issue by batterseapower** [pandas-dev/pandas#28704]
#### Code Sample, a copy-pastable example if possible

These both say `float64`:

```python
pd.DataFrame(np.zeros((0, 0), dtype=int), index=[]).values.dtype
pd.DataFrame(np.zeros((1, 0), dtype=int), index=['A']).values.dtype
```

All of the following examples instead say `int32` as expected:

```python
pd.DataFrame(np.zeros((0, 1), dtype=int), index=[]).values.dtype
pd.DataFrame(np.zeros(0, dtype=int), index=[]).values.dtype
pd.Series(np.zeros(0, dtype=int), index=[]).values.dtype
```

#### Problem description

I think in general it would be a nice property that if `X` is a 2D ndarray, then it should be the case that `pd.DataFrame(X).values.base is X`.

It would be great if pandas could not lose dtype information, because I sometimes wrap 2D arrays in DataFrames before extracting them (via `.values`) for use elsewhere. If that wrapping/unwrapping process changes dtype then it can lead to surprising failures (e.g. I roundtrip a 2D array of indexes and then find that I'm trying to index into another array with a 2D array of floats..)

#### Expected Output

`int32` or `int64` for all examples.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.24.1
pytest: 3.1.2
pip: 19.2.1
setuptools: 39.0.1
Cython: 0.29.6
numpy: 1.17.0
scipy: 1.2.1
pyarrow: 0.13.0
xarray: 0.12.1
IPython: 6.1.0
sphinx: 2.1.1
patsy: 0.4.1
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: 1.3.0.dev0
tables: None
numexpr: 2.6.9
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: 4.3.4
bs4: 4.7.1
html5lib: 0.9999999
sqlalchemy: 1.1.11
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: 0.3.1
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

**Comment by batterseapower** [comment]
Note that you can't even rescue the situation with `astype(int)` on the DataFrame with an erroneous dtype: a DataFrame with zero columns is just stuck as a float64.

**Comment by TomAugspurger** [comment]
I don't think that pandas has a place to represent this. Since each column can have it's own dtype, the "dtype of no column" isn't well defined.

**Comment by jorisvandenbossche** [comment]
Yes, I agree that there is no obvious way to "solve" this. If there are no columns, we cannot store any information about dtypes. In those cases, the "float64" is just the default fallback dtype.

**Comment by batterseapower** [comment]
I'm not completely au-fait with Pandas internals, but I don't think it's quite true that pandas represents a separate dtype for each column, these assertions succeeds:

```
X = np.zeros((3, 2), order='C', dtype='i4')
assert pd.DataFrame(X).values.base is X
assert pd.DataFrame(X).values.flags.c_contiguous
```

i.e. the "int4" dtype associated with columns 0 and 1 is actually shared between the two columns in some sense, because the columns are backed by the same array. (Pandas can't be splitting the 2D array into two columnar slices internally, because then `pd.DataFrame(X).values` would not be C-contiguous.)

With this in mind, I wonder if it possible to represent this in Pandas using an internal 2D array that is shared across 0 columns in just the same way that my `X` above is shared across 2 columns. Not saying that this would be easy to do or worth the effort, but it seems that it might fit into the Pandas internal data model?

**Comment by batterseapower** [comment]
I guess I'm suggesting that the BlockManager could store a single block with 0 columns which would give you somewhere to store the dtype.

**Comment by jorisvandenbossche** [comment]
Personally, that seems adding more complexity than worth it for such a corner case in array - frame roundtrip (you would need to ensure that all of the internal code can handle such a dataframe correctly, that all code that expects blocks with values can handle blocks with no values, that this block gets dropped once other blocks are added, ...).

**Comment by batterseapower** [comment]
The other surprising consequence of this bug is that things like `df.isnull().values` give you a `float64` array instead of a `bool` array specifically when df has 0 columns (the 0 rows case works fine).
